src/Qir/Runtime/public/QubitManager.hpp

```diff
@@ -0,0 +1,220 @@
+        void Release(Qubit* qubitsToRelease, int qubitCountToRelease);
+
+        // Borrow (We treat borrowing as allocation currently)
+        Qubit Borrow();
+        void Borrow(Qubit* qubitsToBorrow, int qubitCountToBorrow);
+        // Return (We treat returning as release currently)
+        void Return(Qubit qubit);
+        void Return(Qubit* qubitsToReturn, int qubitCountToReturn);
+
+        // Disables a given qubit.
+        // Once a qubit is disabled it can never be "enabled" or reallocated.
+        void Disable(Qubit qubit);
+        // Disables a set of given qubits.
+        // Once a qubit is disabled it can never be "enabled" or reallocated.
+        void Disable(Qubit* qubitsToDisable, int qubitCountToDisable);
+
+        bool IsValid(Qubit qubit) const;
+        bool IsDisabled(Qubit qubit) const;
+        bool IsFree(Qubit qubit) const;
+
+        // May be overriden to create a custom Qubit object.
+        // When not overriden, it just stores qubit Id in place of a pointer to a qubit.
+        // id: unique qubit id
+        // Returns a newly instantiated qubit.
+        virtual Qubit CreateQubitObject(QubitIdType id);
+
+        // May be overriden to get a qubit id from a custom qubit object.
+        // Must be overriden if CreateQubitObject is overriden.
+        // When not overriden, it just reinterprets pointer to qubit as a qubit id.
+        // qubit: pointer to QUBIT
+        // Returns id of a qubit pointed to by qubit.
+        virtual QubitIdType QubitToId(Qubit qubit) const;
+
+        // Qubit counts
+        int GetDisabledQubitCount() const { return disabledQubitCount; }
+        int GetAllocatedQubitCount() const { return allocatedQubitCount; }
+        int GetFreeQubitCount() const { return freeQubitCount; }
+
+        bool GetMayExtendCapacity() const { return mayExtendCapacity; }
+        bool GetEncourageReuse() const { return encourageReuse; }
+
+    private:
+        void EnsureCapacity(QubitIdType requestedCapacity);
+
+        QubitIdType AllocateQubitId();
+        void ReleaseQubitId(QubitIdType id);
+        void ChangeStatusToAllocated(QubitIdType id);
+        void Release(QubitIdType id);
+
+        bool IsDisabled(QubitIdType id) const;
+        bool IsExplicitlyAllocated(QubitIdType id) const;
+        bool IsFree(QubitIdType id) const;
+
+        // Configuration Properties
+        bool mayExtendCapacity;
+        bool encourageReuse;
+
+        // State
+        QubitIdType* sharedQubitStatusArray; // Tracks allocation state of all qubits. Stores lists of free qubits.
+        QubitIdType qubitCapacity; // qubitCapacity is always equal to the array size.
+        CRestrictedReuseAreaStack freeQubitsInAreas; // Fresh Free Qubits are located in freeQubitsInAreas[0].FreeQubitsReuseAllowed
+
+        // Counts
+        int disabledQubitCount;
+        int allocatedQubitCount;
+        int freeQubitCount;
```

Comment:
(FYI)
I didn't see any discussion of [this](https://github.com/microsoft/qsharp-runtime/pull/696#discussion_r635652520). To make sure that this particular part is not forgotten, I'll remind, consider using the unsigned types for the unsigned concepts (or I'd love if you could explain me the benefits of the signed types).

Looking forward I also recommend the _fixed-size types_. See details in [Do Not Expose the Non-fixed-size Data Types](https://ms-quantum.visualstudio.com/Quantum%20Program/_git/Quantum-OKRs.wiki?path=%2FMicrosoft-Quantum-Program%2FTechnical%2FEngineering-Specifications%2FCompiler-%26-Runtime%2FQIR-RT-API-Design-Guidelines.md&version=GBrokuzmin%2FqirRtApiGuidelines&_a=preview&anchor=do-not-expose-the-non-fixed-size-data-types) section (the doc is under development).

Reply:
You didn't see any discussion because this github interface ate my replies. I will post them again.

Reply:
Signed vs unsigned is now explained (see QubitIdType comment). This class is not part of the interface, but I think it is beneficial to use int32_t instead of int as it will simplify any wrapper that will expose this functionality. I kept the bool though - it's convenient and can be easily handled in the wrapper if needed.